src/sentry/options/types.py

```diff
@@ -0,0 +1,173 @@
+
+from sentry.utils.yaml import safe_load
+from yaml.parser import ParserError
+from yaml.scanner import ScannerError
+
+__all__ = (
+    'InvalidTypeError',
+    'Any', 'Bool', 'Int', 'Float', 'String', 'Dict', 'Sequence',
+)
+
+
+class InvalidTypeError(TypeError):
+    pass
+
+
+class OptionType(object):
+    """Base OptionType that provides type coersion"""
+    name = ''
+    # Default value to be returned when initializing
+    default = None
+    # Types that do not need to be coerced
+    expected_types = (object,)
+    # Types that are acceptable for coersion
+    compatible_types = (basestring,)
+
+    def __call__(self, value=None):
+        if value is None:
+            return self._default()
+        if self.test(value):
+            return value
+        if isinstance(value, self.compatible_types):
+            rv = self.convert(value)
+            # Make sure convert was able to do the right thing
+            # and give us the type we were expecting
+            if self.test(rv):
+                return rv
+        raise InvalidTypeError('{!r} is not a valid {}'.format(value, repr(self)))
+
+    def convert(self, value):
+        return value
+
+    def _default(self):
+        return self.default
+
+    def test(self, value):
+        """Check if the value needs to be coerced or not"""
+        return isinstance(value, self.expected_types)
+
+    def __repr__(self):
+        return self.name
+
+
+class AnyType(OptionType):
+    """A type that accepts any value and does no coersion"""
+    name = 'any'
+
+
+class BoolType(OptionType):
+    "Coerce a boolean from a string"
+    name = 'boolean'
+    default = False
+    expected_types = (bool,)
+
+    def convert(self, value):
+        if value.lower() in ('y', 'yes', 't', 'true', '1'):
+            return True
+        if value.lower() in ('n', 'no', 'f', 'false', '0'):
+            return False
+
+
+class IntType(OptionType):
+    """Coerce an integer from a string"""
+    name = 'integer'
+    default = 0
+    expected_types = (int,)
+
+    def convert(self, value):
+        try:
+            return int(value)
+        except ValueError:
+            return
+
+
+class FloatType(OptionType):
+    """Coerce a float from a string or integer"""
+    name = 'float'
+    default = 0.0
+    expected_types = (float,)
+    compatible_types = (basestring, float, int)
+
+    def convert(self, value):
+        try:
+            return float(value)
+        except ValueError:
+            return
+
+
+class StringType(OptionType):
+    """String type without any coersion, must be a string"""
+    name = 'string'
+    default = u''
+    expected_types = (basestring,)
+    compatible_types = (basestring,)
+
+
+class DictType(OptionType):
+    """Coerce a dict out of a json/yaml string"""
+    name = 'dictionary'
+    expected_types = (dict,)
+
+    def _default(self):
+        # make sure we create a fresh dict each time
+        return {}
```

Comment:
Maybe `default` should just be a callable all the time, like the `defaultdict` constructor?